### PR TITLE
Trinotate

### PR DESCRIPTION
**Pull request by lecorguille**
A downstream tool of Trinity developped by Trinity 
https://trinotate.github.io/

### CODE REVIEW

**Comment by bgruening** [comment]
There is something wrong with this PR. It has many more commits? :(

**Comment by lecorguille** [comment]
Rrroowww, in January, I will ask you and Anthony an `rebase` extrem training session

Can I delete commit using `git rebase -i 24591d758591dbd0f032fb3732f6e60e5654512a` ?

**Comment by bgruening** [comment]
@lecorguille if you do `git rebase -i` you will get an editor. In this windows edit the first column and change `pick` to `d` for your specific commit. Moreover, read the docu in this editor window it's quite useful. Save your results and close the windows.

**Comment by mvdbeek** [comment]
Also delete in rebase is "relatively" new, make sure you're on git 2.x.
Otherwise you can always create a new branch and cherry-pick the commits you want.

**Comment by lecorguille** [comment]
lol

**Comment by lecorguille** [comment]
Yes :)
I was afraid to reboot Internet

**Comment by mvdbeek** [comment]
The test data is huge, any way you can trim this down a bit?

**Comment by lecorguille** [comment]
I took those provide by Trinotate developers
But sure, I will do that ... by not immediately ⏲

**Comment by lecorguille** [comment]
Passed but it seems that there were not real tests

File: tools/trinotate/trinotate.xml [hunk 1]
```diff
@@ -0,0 +1,186 @@
+<tool id="trinotate" name="Trinotate" version="0.1.0">
+    <requirements>
+        <requirement type="package" version="3.0.1">trinotate</requirement>
+    </requirements>
+    <stdio>
+        <exit_code range="1:" />
+    </stdio>
+    <command><![CDATA[
+wget --no-verbose "https://data.broadinstitute.org/Trinity/Trinotate_v3_RESOURCES/Trinotate_v3.sqlite.gz" -O Trinotate.sqlite.gz &&
```
Comment (bgruening) [inline tools/trinotate/trinotate.xml]: This means you would need a requirement on gnu_wget. But I would rather see this as an input and state in the tool help the URL to download this sqlite file. The reason is that not all clusters and nodes do have internet connection.

File: tools/trinotate/trinotate.xml [hunk 2]
```diff
@@ -0,0 +1,186 @@
+<tool id="trinotate" name="Trinotate" version="0.1.0">
```
Comment (bgruening) [inline tools/trinotate/trinotate.xml]: Can we have a tool version of 3.0.1

File: tools/trinotate/trinotate.xml [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+    </requirements>
+    <stdio>
+        <exit_code range="1:" />
+    </stdio>
+    <command><![CDATA[
+wget --no-verbose "https://data.broadinstitute.org/Trinity/Trinotate_v3_RESOURCES/Trinotate_v3.sqlite.gz" -O Trinotate.sqlite.gz &&
+gunzip Trinotate.sqlite.gz &&
+
+Trinotate Trinotate.sqlite init --gene_trans_map "$gene_trans_map" --transcript_fasta $trinity --transdecoder_pep "$transdecoder" &&
+Trinotate Trinotate.sqlite LOAD_swissprot_blastp "$blastp" &&
```
Comment (bgruening) [inline tools/trinotate/trinotate.xml]: Singel quotes are recommended.

File: tools/trinotate/trinotate.xml [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+                <option value="SGC">SGC : sequence gathering cutoff</option>
+                <option value="STC">STC : sequence trusted cutoff</option>
+            </param>
+            <param argument="--incl_pep" type="boolean" truevalue="--incl_pep" falsevalue="" checked="false" label="Add the protein sequence data in the tab delimited report" />
+            <param argument="--incl_trans" type="boolean" truevalue="--incl_trans" falsevalue="" checked="false"  label="Add the transcript sequence data in the tab delimited report" />
+        </section>
+        <param name="keep_sqlite" type="boolean" truevalue="true" falsevalue="false" checked="false" label="Keep the sqlite database file: Trinotate.sqlite" help="Can be used for visualization purpose" />
+    </inputs>
+    <outputs>
+        <data name="trinotate_out" format="tabular" label="Trinotate output" />
```
Comment (bgruening) [inline tools/trinotate/trinotate.xml]: You could add metadata to indicate the column headers as in the DESeq2 wrapper.

File: tools/trinotate/trinotate.xml [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+            Trinotate version 3.0.1
+        ]]>
+    </version_command>
+    <command><![CDATA[
+
+#if str($cond_download.select_download) == "yes":
+    wget --no-verbose 'https://data.broadinstitute.org/Trinity/Trinotate_v3_RESOURCES/Trinotate_v3.sqlite.gz' -O Trinotate.sqlite.gz &&
+    gunzip Trinotate.sqlite.gz &&
+#else:
+    ln -s $cond_download.trinotate_sqlite_in Trinotate.sqlite
```
Comment (bgruening) [inline tools/trinotate/trinotate.xml]: quote ...

File: tools/trinotate/trinotate.xml (hunk 2 above)
Comment (bgruening) [inline tools/trinotate/trinotate.xml]: ping

File: tools/trinotate/trinotate.xml [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+#if $optional.signalp:
+    Trinotate Trinotate.sqlite LOAD_signalp '$optional.signalp' &&
+#end if
+
+Trinotate Trinotate.sqlite report -E $report_option.E --pfam_cutoff $report_option.pfam_cutoff $report_option.incl_pep $report_option.incl_trans > '$trinotate_out';
+
+    ]]></command>
+    <inputs>
+        <param name="trinity" type="data" format="fasta" optional="false" label="Transcripts" help="Trinity output: Assembled Transcripts" />
+        <param name="transdecoder" type="data" format="fasta" optional="false" label="Peptides" help="Trinotate output: transdecoder_pep" />
```
Comment (abretaud) [inline tools/trinotate/trinotate.xml]: Do you mean "transdecoder output"?

File: tools/trinotate/trinotate.xml [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+        <param name="gene_trans_map" type="data" format="tabular" optional="false" label="Genes to transcripts map" help="Genes to transcripts map output. Containing correspondance between gene ids and transcript ids based on the name of transcripts assembled by Trinity" />
```
Comment (abretaud) [inline tools/trinotate/trinotate.xml]: Is it required? For trinity diff expression scripts it's only required if the input transcriptome fasta was not generated by the latest trinity version

File: tools/trinotate/trinotate.xml (hunk 2 above)
Comment (lecorguille) [inline tools/trinotate/trinotate.xml]: @bgruening 

about your first review, is `<version_command><![CDATA[ Trinotate version 3.0.1 ]]>` or did you expect something else

For the last one, is it just ping the version number to `1.0.0`?

File: tools/trinotate/trinotate.xml (hunk 6 above)
Comment (lecorguille) [inline tools/trinotate/trinotate.xml]: Indeed

File: tools/trinotate/trinotate.xml (hunk 7 above)
Comment (lecorguille) [inline tools/trinotate/trinotate.xml]: It seems so
I tried without by it's require
```
###########################################################################
#
# Required:
#
#  --sqlite <string>                  Trinotate sqlite database file
#
#  --gene_trans_map <string>          text file indicating  "gene<tab>trans"
#                                      for each trans-gene relationship.
#
#  --transcript_fasta <string>        transcripts fasta file
#
#  --transdecoder_pep <string>        transcdecoder-generated peptide file
#
# Optional:
#
#  --bulk_load                        for faster loading
#
###########################################################################
```

Comment (lecorguille) [inline tools/trinotate/trinotate.xml]: But I have change to optional blast* and hmmscan

**Comment by abretaud** [comment]
For the version, you'd better use a @WRAPPER_VERSION@ token (as in trinity for example).
Ok for the gene_map param, maybe we could (one day!) launch it automatically when the assembly comes from trinity?

**Comment by bgruening** [comment]
:+1: by me. @abretaud has the last call :)

File: tools/trinotate/trinotate.xml [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+            <param argument="--incl_pep" type="boolean" truevalue="--incl_pep" falsevalue="" checked="false" label="Add the protein sequence data in the tab delimited report" />
+            <param argument="--incl_trans" type="boolean" truevalue="--incl_trans" falsevalue="" checked="false"  label="Add the transcript sequence data in the tab delimited report" />
+        </section>
+        <param name="keep_sqlite" type="boolean" truevalue="true" falsevalue="false" checked="false" label="Keep the sqlite database file: Trinotate.sqlite" help="Can be used for visualization purpose" />
+    </inputs>
+    <outputs>
+        <data name="trinotate_out" format="tabular" label="Trinotate output" >
+            <actions>
+                <action name="column_names" type="metadata" default="gene_id,transcript_id,sprot_Top_BLASTX_hit,RNAMMER,prot_id,prot_coords,sprot_Top_BLASTP_hit,custom_pombe_pep_BLASTX,custom_pombe_pep_BLASTP,Pfam,SignalP,TmHMM,eggnog,Kegg,gene_ontology_blast,gene_ontology_pfam,transcript,peptide
+" />
```
Comment (abretaud) [inline tools/trinotate/trinotate.xml]: there shouldn't be a line break here I think

**Comment by abretaud** [comment]
Looks good!
I had not seen there is a single wrapper, so you can forget the @WRAPPER_VERSION@ thing.
You will hate me, but it would be cool to rebase/squash some commits ? Otherwise the huge test-data will stay in the git history

**Comment by bgruening** [comment]
@lecorguille we can also squash the commits during merge, but then your commit stats drop ;) If you care about this you need to rebase it.

**Comment by lecorguille** [comment]
Damn!
My christmas bonus is computed with my number of commit 😭 

No problemo

**Comment by bgruening** [comment]
You can choose to rebase yourself ;)
Alternatively, you get a X-mas bonus from the IUC!

**Comment by lecorguille** [comment]
You know how I fond of `git rebase`
I definitly choose the alternative 👍

**Comment by bgruening** [comment]
Thanks @lecorguille!

**Comment by lecorguille** [comment]
Thanks to you for your reviews

**Comment by lecorguille** [comment]
Heu ...  I guess I should have include a .shed.yml ...

**Pull request merged**